### DISCUSSION TITLE
I am getting File "src/pymssql/_pymssql.pyx", line 479, in pymssql._pymssql.Cursor.execute pymssql._pymssql.OperationalError: (3971, b'The server failed to resume the transaction. Desc:4900000003.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')

### DISCUSSION THREAD

**Discussion by manmayaray** [sqlalchemy/sqlalchemy#11809] [Usage Questions]
### Describe the use case

File "src/pymssql/_pymssql.pyx", line 479, in pymssql._pymssql.Cursor.execute
pymssql._pymssql.OperationalError: (3971, b'The server failed to resume the transaction. Desc:4900000003.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')

### Databases / Backends / Drivers targeted

mssql server 16, SQL Alchemy 1.4, Python 3.8 and pymysql 2.27 

### Example Use
```
Traceback (most recent call last):
  File "src/pymssql/_pymssql.pyx", line 459, in pymssql._pymssql.Cursor.execute
  File "src/pymssql/_mssql.pyx", line 1087, in pymssql._mssql.MSSQLConnection.execute_query
  File "src/pymssql/_mssql.pyx", line 1118, in pymssql._mssql.MSSQLConnection.execute_query
  File "src/pymssql/_mssql.pyx", line 1251, in pymssql._mssql.MSSQLConnection.format_and_run_query
  File "src/pymssql/_mssql.pyx", line 1789, in pymssql._mssql.check_cancel_and_raise
  File "src/pymssql/_mssql.pyx", line 1835, in pymssql._mssql.raise_MSSQLDatabaseException
pymssql._mssql.MSSQLDatabaseException: (3971, b'The server failed to resume the transaction. Desc:4900000003.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3256, in _wrap_pool_connect
    return fn()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 901, in _checkout
    result = pool._dialect.do_ping(fairy.dbapi_connection)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 699, in do_ping
    cursor.execute(self._dialect_specific_select_one)
  File "src/pymssql/_pymssql.pyx", line 479, in pymssql._pymssql.Cursor.execute
pymssql._pymssql.OperationalError: (3971, b'The server failed to resume the transaction. Desc:4900000003.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')
 
The above exception was the direct cause of the following exception:
 
Traceback (most recent call last):
  File "/home/op_general/icici_medical_api_V2/portal/routes/file_upload/controller.py", line 150, in post
    header = APIHeader.query.filter_by(claim_header_id=header_id).first()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2810, in first
    return self.limit(1)._iter().first()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2894, in _iter
    result = self.session.execute(
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1691, in execute
    conn = self._connection_for_bind(bind)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1532, in _connection_for_bind
    return self._transaction._connection_for_bind(
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 747, in _connection_for_bind
    conn = bind.connect()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3210, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3289, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3259, in _wrap_pool_connect
    Connection._handle_dbapi_exception_noconnection(
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2106, in _handle_dbapi_exception_noconnection
    util.raise_(
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3256, in _wrap_pool_connect
    return fn()
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 901, in _checkout
    result = pool._dialect.do_ping(fairy.dbapi_connection)
  File "/home/op_general/miniconda/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 699, in do_ping
    cursor.execute(self._dialect_specific_select_one)
  File "src/pymssql/_pymssql.pyx", line 479, in pymssql._pymssql.Cursor.execute
sqlalchemy.exc.OperationalError: (pymssql._pymssql.OperationalError) (3971, b'The server failed to resume the transaction. Desc:4900000003.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')
```
### Additional context
```py
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.pool import QueuePool
from portal import create_app
db = SQLAlchemy()



def init_app(app):
    db_connection_string = app.config["DB_CONNECTION_STRING"]
    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
    app.config['SQLALCHEMY_BINDS'] = {
        'db': db_connection_string
    }
    app.config['SQLALCHEMY_DATABASE_URI']=db_connection_string
    app.config['SQLALCHEMY_POOL_RECYCLE'] = 299
    app.config['SQLALCHEMY_ENGINE_OPTIONS'] = {
        "poolclass": QueuePool,
        "pool_size": 1000,
        "pool_pre_ping": True,
        "pool_recycle": 300,
        "max_overflow": 1000,
```

**Comment by manmayaray** [comment]
I am really struggling for past few days with this

**Comment by CaselIT** [comment]
Hi

Seems to be a connection error from the server. Not really sure what you expect sqlalchemy to do here

cc also @gordthompson

**Comment by manmayaray** [comment]
Hi, 
 We are getting this error from SQL server when we are sending more than 10 to 15 requests. Sometimes mo issues at all and some times issues are there. I have added sqlalchmey engine parameters to suggest if anything is needed to change to avoid this error.

def init_app(app):
    db_connection_string = app.config["DB_CONNECTION_STRING"]
    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
    app.config['SQLALCHEMY_BINDS'] = {
        'db': db_connection_string
    }
    app.config['SQLALCHEMY_DATABASE_URI']=db_connection_string
    app.config['SQLALCHEMY_POOL_RECYCLE'] = 299
    app.config['SQLALCHEMY_ENGINE_OPTIONS'] = {
        "poolclass": QueuePool,
        "pool_size": 1000,
        "pool_pre_ping": True,
        "pool_recycle": 300,
        "max_overflow": 1000,

**Reply by gordthompson** [reply]
cross-posted here: https://github.com/pymssql/pymssql/issues/901

**Comment by zzzeek** [comment]
@CaselIT @gordthompson the stack trace shows this happening in do_ping which at least indicates the pymssql ping routine is not recovering from a connectivity error, we would likely want to add this as a disconnect exception for pymssql

@manmayaray try using NullPool for the moment and see if errors resolve, since you have connections that are in an errored state being returned to the pool

**Reply by CaselIT** [reply]
yes, I was wondering the same. we seem to have message 20017 here https://github.com/sqlalchemy/sqlalchemy/blob/22cbc7dcb48c946dda66704797665289965eb22e/lib/sqlalchemy/dialects/mssql/pymssql.py#L95-L110, maybe also 20018 should be treated the same way?

**Reply by zzzeek** [reply]
probably, it looks like a bad connection in their pool, so sure

**Reply by CaselIT** [reply]
from here it seems to be a general "query rejected" error https://github.com/tds-fdw/tds_fdw/issues/105#issuecomment-259769194

probably in this case if it reject `select 1` it's safe to consider it a disconnect error. Not sure in general if it would be the case though

**Reply by zzzeek** [reply]
my only concern is this is a reason to reject the single connection but not the entire pool.  im not sure if a failed ping means the whole pool gets invalidated

**Comment by manmayaray** [comment]
ok, so suggestion is to change QueuePoll to Null pool?

**Reply by zzzeek** [reply]
well just try it as part of a fact finding attempt

**Reply by manmayaray** [reply]
we changed and did testing in lower env , it seems to be working
tried for 5, 10 and 20 concurrent requests system didnt fail for NullPool ..however few hours back we had tried Queuepool and it had failed for the same requests trend

**Reply by manmayaray** [reply]
We will migrate the changes to Prod on Monday and observer...Just couple of question, is it possible after working for few days , this my stop working and we can still see issue
s

**Reply by manmayaray** [reply]
Also is there any impact to performance using NullPool in place of Queue pool, in the production  we get usually 150 transactions per hours, where multiple file processing, data extraction and validation works, right now from 12 to 5 close to 60 to 70 transactions going to exception due to this error.

**Reply by zzzeek** [reply]
Nullpool will perform worse if you're going for tens or hundreds of transactions per second.

Using nullpool here is a workaround.  In your app, before you get the error that you're reporting here, something else is going wrong on your connection that is causing an unresolvable transaction to be left open.  You probably see some kind of error on a previous request.

**Reply by manmayaray** [reply]
> Nullpool will perform worse if you're going for tens or hundreds of transactions per second.
> 
> Using nullpool here is a workaround. In your app, before you get the error that you're reporting here, something else is going wrong on your connection that is causing an unresolvable transaction to be left open. You probably see some kind of error on a previous request.

Hi @zzzeek , The only parameter is being used in the connection engine is  pool_pre_ping=True, while going through sqlalchmey it is stating that , adding this will not ensure that the connection pool is ,f the ping / error check determines that the connection is not usable, the connection will be immediately recycled, and all other pooled connections older than the current time are invalidated, so that the next time they are checked out, they will also be recycled before use

**Reply by zzzeek** [reply]
correct.  preping is running a query on the connection, and ideally it would determine the connection is unusuable, and that is a bug.  however, the manner in which your connection is unusable, is the result of a previous transaction failing in some way that we are unfamiliar with.   If you were to fix that problem, the issue would also go away.   It's also important that you try to fix that problem, because if you are getting lots of these errors, each time pre_ping fails, it will *recycle the whole pool*, which means your performance will not be much better than NullPool if this error occurs constantly.

that is, there's things to fix all around but it would be best if you could isolate the conditions that cause your driver to emit this message in the first place.  we can't do that here since we dont have your code.

**Comment by manmayaray** [comment]
Hello, on further investigation I found the link where it is mentioned its a bug in SQLAlchemy. but unable to open the link
 https://stackoverflow.com/questions/5549940/sqlalchemy-connections-arent-closed-when-pool-is-overflowed?rq=4 http://hg.sqlalchemy.org/sqlalchemy/rev/aff95843c12f#l2.17

**Reply by CaselIT** [reply]
It's unlikely that a 13 year old discussion has any relevance today

**Reply by gordthompson** [reply]
That bug was fixed back in 2011.

https://github.com/sqlalchemy/sqlalchemy/commit/a294df9748f1314a2db7de5bc39d091854d0fec3

**Comment by zzzeek** [comment]
#11822 will be in release 2.0.33

**Comment by manmayaray** [comment]
Hi team, I have a query will adding Autocommit help to recover from this problem? Anyways we have added dp_api exception to gracefully handle the exception.

**Comment by manmayaray** [comment]
Hi this is the block which is causing the issue and I have added the recent exception handling
```
Try
 db.session.add(header)
            db.session.commit()
            db.session.close()
            LOG.info(f"Case_start: Successfully created new claim with header_id: {header_id}")
            return {"success": True, "data": {"XXX": header_id}, "errors": ""}

 except OperationalError as e:
            db.session.rollback()
            db.session.close()
            LOG.error(f"Case_start: OperationalError, error: {str(e)}", exc_info=True)
            LOG.error("Full traceback: " + traceback.format_exc())
            raise InternalServerError("Database is temporarily unavailable, please try again later.")
        except DBAPIError as e:
            db.session.rollback()
            db.session.close()
            LOG.error(f"Case_start: DBAPIError, error: {str(e)}", exc_info=True)
            LOG.error("Full traceback: " + traceback.format_exc())
            raise InternalServerError("Database is temporarily unavailable, please try again later.")
        except SQLAlchemyError as e:
            db.session.rollback()
            db.session.close()
            LOG.error(f"Case_start: General SQLAlchemyError, error: {str(e)}", exc_info=True)
            LOG.error("Full traceback: " + traceback.format_exc())
            raise InternalServerError("An unexpected database error occurred.")
        except Exception as e:
            db.session.rollback()
            db.session.close()
            LOG.error(f"Case_start: Error while processing new claim, error: {str(e)}", exc_info=True)
            LOG.error("Full traceback: " + traceback.format_exc())
            raise InternalServerError("Error while processing new claim")

**Comment by manmayaray** [comment]
Another point will SQLAlchmey 2.0, we will not face this issue. right now we are doing retry and few places exceptions with Operationar, DB_API and SQAlchemry error

**Comment by manmayaray** [comment]
Adding exception etc is not helping, more than 15 sessions system is not able to accept.Do we have alternate approach to solve this?Hourly basis close to 20 transactions out of 140+ are failing from peak time 11 to 5 PM
Is it possible to release sessions explicitly in exception
 handling

**Comment by manmayaray** [comment]
@CaselIT @zeeeeeb , should I remove pre_ping ??, pls help
